sympy/tensor/array/ndim_array.py

```diff
@@ -329,7 +329,7 @@ def __sub__(self, other):
         from sympy.tensor.array.arrayop import Flatten
 
         if not isinstance(other, NDimArray):
-            raise TypeError(str(other))
+            raise NotImplementedError('Unsupported operand type(s) for - : ' + str(type(other)) + ' and ' + str(type(self)))
```

Comment:
Raising `NotImplementedError` is not the same as returning `NotImplemented`.

Functions like `__sub__` should check for the expected types and return `NotImplemented` if they don't recognise the type. That makes it possible for another classes to define `__rsub__` and then interact with this one.

Reply:
I tried NotImplemented first but that gave flake error with a msg that NotImplementedError should be used istead of NotImplemented

Reply:
It should be `return NotImplemented` rather than `raise NotImplemented`.

Reply:
Oh  yes my bad , I didn't realize I was raising instead of returning